test.py

```diff
@@ -191,22 +262,112 @@ def pattern(self) -> str:
+        """Run tests for this suite"""
+        runner_id = self._runner_newid()    # id for this running task for this specific suite
+        # logger for suite runner, tests will change cluster logger when they use it
+        loggerPrefix = f"{self.mode}/{self.name}-{runner_id}"
+        logger = LogPrefixAdapter(logging.getLogger(loggerPrefix), {'prefix': loggerPrefix})
+        # logger.info("starting test runner")
+        failed: int = 0   # failed tests
+        test_args = await self._run_init(logger)                   # runner setup (e.g. cluster)
+
+        # NOTE: an already running worker may complete this suite before memory is
+        #       available, so check if there are tests remaining to run
+        # Wait for minimum available memory before starting a runner for this suite
+        # print(f"XXX {worker_name} {runner_id} TestSuite.run() {self.mode}/{self.name} requesting INITIAL memory {self.min_memory}")
+        if not await memory_manager.reserve(self.min_memory, lambda: self.remaining > 0
+                                            , debugmsg=f"{worker_name}-{runner_id} {self.mode}/{self.name} INITIAL"  # XXX debug
+                                            ):
+            logger.warning("Suite %s/%s runner %s not enough memory", self.mode, self.name, runner_id)
+            return
+
+        while True:
+            # Get next test
+            async with lock:
+                if self.remaining:
+                    # print(f"XXX {worker_name} {runner_id} TestSuite.run() remaining={self.remaining} - len {len(self.tests)} - picking {len(self.tests) - self.remaining}")
+                    test = self.tests[len(self.tests) - self.remaining]    # pick next test
+                    logger.info("Suite %s - %s runner %s scheduling test %s  REMAINING=%s",
+                                   self.mode, self.name, runner_id, test.name, self.remaining)
+                    self.remaining -= 1
+                    self.remaining_first -= test.run_first
+                    # print(f"XXX {self.mode}/{self.name} TestSuite.run() {worker_name} {runner_id} scheduling test {test.name} run_first {test.run_first} remaining {self.remaining}")
+                    logger.info("Suite %s - %s runner %s scheduling test %s",
+                                self.mode, self.name, runner_id, test.name)
+                else:
+                    # print(f"XXX TestSuite.run() {worker_name} {runner_id} {self.mode}/{self.name} NO MORE TESTS")
+                    break   # no more tests for this suite
+
+            # Reserve extra memory for this test (if non-zero)
+            # XXX shortname is not really short name for Boost
+            #print(f"XXX TestSuite.run() {worker_name} {runner_id} {self.mode}/{self.name} test {test.shortname} going to reserve test memory {self._test_memory(test.shortname)}")
+            if not await memory_manager.reserve(self._test_memory(test.shortname)
```

Comment:
Why are you reserving here on top of min_memory reserved earlier? Shouldn't you only add the delta between min_memory and test_memory?

Reply:
The test extra memory is a delta over the min memory. Perhaps I should rename it in the config, but the comment says so.